plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java

```diff
@@ -310,6 +321,18 @@ protected String getTableSchemaName(ResultSet resultSet)
 
             case Types.DATE:
                 return Optional.of(dateColumnMapping());
+
+            case Types.TIMESTAMP:
+                // TODO support higher precisions
+                TimestampType timestampType = TIMESTAMP_MILLIS;
+                return Optional.of(ColumnMapping.longMapping(
+                        timestampType,
+                        (resultSet, columnIndex) -> {
+                            Timestamp timestamp = resultSet.getTimestamp(columnIndex);
+                            Instant instant = timestamp.toLocalDateTime().atZone(UTC).toInstant();
+                            return toPrestoTimestamp(timestampType, LocalDateTime.ofEpochSecond(instant.getEpochSecond(), Math.toIntExact(round(instant.getNano(), 9 - timestampType.getPrecision())), UTC));
```

Comment:
The _default_ read function is not correct -- it should either round OR fail, but should not return incorrect data silently as it does today.

Reply:
I don't think it returns incorrect data - a VerifyException is thrown if the value and the precision don't match.

The issue here was that the value returned from MySQL has precision != 3 and hence the `verify` fails. It's our fault as callers of the code to not provide the correct `TimestampType` (we pass `TIMESTAMP_MILLIS` always).

Should be addressed as part of https://github.com/trinodb/trino/issues/6910.

Reply:
> I don't think it returns incorrect data - a VerifyException is thrown if the value and the precision don't match.

Right. That's a new addition (#6654), and so #6852 reported a different stacktrace than one we would get today.

Still, we could add an opt-in mode where the value is rounded (can be eg `boolean round`)
This should be used by all connectors which use TIMESTAMP_MILLIS (not only MySQL, but also the legacy default mappings (#6607) and maybe some others.

Reply:
Note that we need to propagate the `boolean` through `legacyToPrestoType` -> `timestampColumnMapping` -> `toPrestoTimestamp`.

I think it's more worthwhile to add the `boolean` (or a method with more meaningful name) to `toPrestoTimestamp` itself and add explicit column mapping for timestamps to connectors which are using `legacyToPrestoType`.

I would like to do this in a separate PR though.

Reply:
> Note that we need to propagate the boolean through legacyToPrestoType -> timestampColumnMapping -> toPrestoTimestamp.

`legacyToPrestoType` should not take this boolean. This is the place where we make decision to map to TS_MILLIS. 
But yes, you'd need to pass a boolean to `timestampColumnMappingUsingSqlTimestamp` and on.